update security concept and examples (#9648)
istio/istio.io: content/en/docs/concepts/security/index.md

the following benefits:

- Workload-to-workload and end-user-to-workload authorization.

> - A Simple API: it includes a single [`AuthorizationPolicy` CRD](/docs/reference/config/security/authorization-policy/),

Reviewer: ```suggestion
- A simple API: it includes a single [`AuthorizationPolicy` CRD](/docs/reference/config/security/authorization-policy/),
```

  which is easy to use and maintain.
- Flexible semantics: operators can define custom conditions on Istio attributes, and use CUSTOM, DENY and ALLOW actions.

> - High performance: Istio authorization (`ALLOW` and `DENY`) is enforced natively on Envoy.

Author: This is called out for `ALLOW` and `DENY`, which is only enforced on Envoy (not configurable).

> - High performance: Istio authorization of DENY and ALLOW action is enforced natively on Envoy.

Reviewer: Technically correct but at the expense of it being an easy-to-read overview.  Is it OK in a concepts page to leave it as it was?

Author: Updated to make this shorter, I called out `ALLOW` and `DENY` since the `CUSTOM` is not enforced in local Envoy.

> - High compatibility: supports gRPC, HTTP, HTTPS and HTTP2 natively, as well as any plain TCP protocols.

Reviewer: ```suggestion
- High compatibility: supports gRPC, HTTP, HTTPS and HTTP/2 natively, as well as any plain TCP protocols.
```

### Implicit enablement

You don't need to explicitly enable Istio's authorization features. It's ready for use out-of-box and you just need to

> apply an authorization policy to the workloads to enforce access control.

Reviewer 2: ```suggestion
You don't need to explicitly enable Istio's authorization features; they are available after installation. To enforce access control to your workloads, you apply an authorization policy.
```

> For workloads without authorization policies applied, Istio doesn't enforce access control allowing all requests.

Reviewer 2: ```suggestion
For workloads without authorization policies applied, Istio allows all requests.
```

Authorization policies support `CUSTOM`, `DENY` and `ALLOW` actions and there are 3 layers of checks: `CUSTOM`, `DENY` and `ALLOW`.

> Each can be applied independently. If none of the layer rejects, the request passes through.

Reviewer 2: ```suggestion
Authorization policies support `CUSTOM`, `DENY` and `ALLOW` actions. You can apply multiple policies, each with a different action, as needed to secure access to your workloads. Istio checks for matching policies in layers, in this order: `CUSTOM`, `DENY`, and then `ALLOW`. For each type of action, Istio first checks if there is a policy with  the action applied, and then checks if the request matches the policy's specification. If a request doesn't match a policy in one of the layers, the check continues to the next layer.
```

### Implicit enablement

> You don't need to explicitly enable Istio's authorization features; they are available after installation.

Reviewer 3: Does some of the content belong to the security tasks? I feel the concept page is pretty long.

To enforce access control to your workloads, you apply an authorization policy.

For workloads without authorization policies applied, Istio allows all requests.

Authorization policies support `CUSTOM`, `DENY` and `ALLOW` actions. You can apply multiple policies, each with a

> different action, as needed to secure access to your workloads.

Reviewer: Many fewer people will use CUSTOM, right?  I'd stay with ALLOW, DENY and CUSTOM in that order

Author: updated.

#### `allow-nothing`, `deny-all` and `allow-all` policy

The following example shows an `ALLOW` policy that matches nothing. If there are no other `ALLOW` policies, requests
will always be denied because of the "deny by default" behavior when there are `ALLOW` policies applied but none matched.

> Note the "deny by default" behavior applies only to the `ALLOW` policy and not other policies.

Reviewer 2: Apologies, I don't know GitHub that well. Anyway, please change the allow-nothing example to the following:

```
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-nothing
spec:
  action: ALLOW
  {}
```

The existing comment really threw me off. I didn't realize that it was talking about the action field. I think it's better to simply add the action here since we say it's a best practice to always include the action.

Author: I updated the example and its comments, it is for the `rules` field.

The following example shows an `ALLOW` policy that matches nothing. If there are no other `ALLOW` policies, requests
will always be denied because of the "deny by default" behavior.

> Note the "deny by default" behavior applies only if the workload has at least one authorization policy with 'ALLOW' action.

Reviewer: ```suggestion
Note the "deny by default" behavior applies only if the workload has at least one authorization policy with the `ALLOW` action.
```

### Dependency on mutual TLS

Istio uses mutual TLS to securely pass some information from the client to the server. Mutual TLS must be enabled before
using any of the following fields in the authorization policy:

- the `principals` and `notPrincipals` field under the `source` section
- the `namespaces` and `notNamespaces` field under the `source` section
- the `source.principal` custom condition
- the `source.namespace` custom condition

Note it is strongly recommended to always use these fields with strict mutual TLS mode in the `PeerAuthentication` to avoid
potential unexpected requests rejection or policy bypass when plain text traffic is used with the permissive mutual TLS mode.

Check the [security advisory](/news/security/istio-security-2021-004) for more details and alternatives if you cannot enable
strict mutual TLS mode.

> Mutual TLS is not required if you don't use any of the above fields in the authorization policy.

Reviewer: Not required for this, but still recommended?  Could this be worded differently?

Author: With the new update, I think we can just remove this sentence.
